geo/src/algorithm/kmeans/mod.rs

```diff
@@ -0,0 +1,1213 @@
+//! - `Ok(labels)`: A vector of cluster labels (0-based indices), one for each input point
+//! - `Err(KMeansError::InvalidK)`: If `k` is 0 or greater than the number of points
+//! - `Err(KMeansError::EmptyCluster)`: If an empty cluster cannot be recovered during iteration
+//! - `Err(KMeansError::InitializationFailed)`: If k-means++ initialisation fails due to degenerate data (e.g., all points at same location, NaN/infinite coordinates)
+//! - `Err(KMeansError::MaxIterationsReached)`: If max_iter is reached before convergence. The error contains the partial result with convergence metadata.
+//!
+//! # Algorithm
+//!
+//! The algorithm works as follows:
+//! 1. **Initialisation**: Uses k-means++ for better initial centroids
+//! 2. **Assignment**: Assign each point to the cluster with the nearest centroid using Hamerly's triangle inequality bounds
+//!    - Maintains upper and lower bounds on distances to avoid redundant calculations
+//!    - Only computes actual distances when bounds don't rule out assignment changes
+//! 3. **Update**: Recalculate centroids as the mean of all points in each cluster
+//! 4. **Bounds update**: Update distance bounds based on centroid movement using triangle inequality
+//! 5. **Convergence check**: Stop if:
+//!    - Centroids have moved less than the tolerance threshold, OR
+//!    - No points have changed cluster assignment, OR
+//!    - Maximum iterations reached
+//! 6. Repeat steps 2-5 until convergence
+//! 7. **max_radius constraint** (if specified): Split any clusters with radius exceeding max_radius
+//!
+//! # Notes
+//!
+//! Empty clusters may rarely occur during iteration if all points are reassigned away from a centroid.
+//! When this happens, the algorithm attempts to recover by reassigning the farthest point from its
+//! current centroid to the empty cluster (following scikit-learn's approach). If recovery fails,
+//! the function returns an error.
+//!
+//! ## Complexity
+//!
+//! - **Initialization (k-means++)**: O(n·k²)
+//! - **Per iteration**: O(k²) centroid distances + O(n) for point assignments with Hamerly pruning
+//! - **Overall**: Effectively O(i·n) for moderate k, where i is iterations to convergence
+//!
+//! For large datasets with _k_ > 50, consider hierarchical or density-based clustering (DBSCAN).
+//!
+//! # Examples
+//!
+//! ## Basic clustering with MultiPoint
+//!
+//! ```
+//! use geo::{KMeans, MultiPoint, point};
+//!
+//! let points = MultiPoint::new(vec![
+//!     point!(x: 0.0, y: 0.0),
+//!     point!(x: 1.0, y: 0.0),
+//!     point!(x: 0.0, y: 1.0),
+//!     point!(x: 10.0, y: 10.0),
+//!     point!(x: 11.0, y: 10.0),
+//!     point!(x: 10.0, y: 11.0),
+//! ]);
+//!
+//! let labels = points.kmeans(2).unwrap();
+//!
+//! // Two clusters should be found
+//! let cluster_0_count = labels.iter().filter(|&l| *l == 0).count();
+//! let cluster_1_count = labels.iter().filter(|&l| *l == 1).count();
+//! assert_eq!(cluster_0_count + cluster_1_count, 6);
```

Comment:
I wonder if it'd be useful to add some kind of helper that iterates through points by cluster (as followup work, not in this PR!)